signature/cose/envelope.go

```diff
@@ -0,0 +1,587 @@
+	coseSigner, err := cose.NewSigner(signer.alg, signer.cryptoSigner)
+	if err != nil {
+		return nil, err
+	}
+	return coseSigner.Sign(rand, payload)
+}
+
+// CertificateChain implements signer interface.
+func (signer *localSigner) CertificateChain() []*x509.Certificate {
+	return signer.certs
+}
+
+type envelope struct {
+	base *cose.Sign1Message
+}
+
+// NewEnvelope initializes an empty COSE signature envelope.
+func NewEnvelope() signature.Envelope {
+	return &base.Envelope{
+		Envelope: &envelope{},
+	}
+}
+
+// ParseEnvelope parses envelopeBytes to a COSE signature envelope.
+func ParseEnvelope(envelopeBytes []byte) (signature.Envelope, error) {
+	var msg cose.Sign1Message
+	if err := msg.UnmarshalCBOR(envelopeBytes); err != nil {
+		return nil, &signature.InvalidSignatureError{Msg: err.Error()}
+	}
+	return &base.Envelope{
+		Envelope: &envelope{
+			base: &msg,
+		},
+		Raw: envelopeBytes,
+	}, nil
+}
+
+// Sign implements signature.Envelope interface.
+// On success, this function returns the COSE signature envelope byte slice.
+func (e *envelope) Sign(req *signature.SignRequest) ([]byte, error) {
+	// get built-in signer from go-cose or remote signer based on req.Signer
+	signer, err := getSigner(req.Signer)
+	if err != nil {
+		return nil, &signature.InvalidSignRequestError{Msg: err.Error()}
+	}
+
+	// prepare COSE_Sign1 message
+	msg := cose.NewSign1Message()
+
+	// generate protected headers of COSE envelope
+	msg.Headers.Protected.SetAlgorithm(signer.Algorithm())
+	if err := generateProtectedHeaders(req, msg.Headers.Protected); err != nil {
+		return nil, &signature.InvalidSignRequestError{Msg: err.Error()}
+	}
+
+	// generate payload of COSE envelope
+	msg.Headers.Protected[cose.HeaderLabelContentType] = req.Payload.ContentType
+	msg.Payload = req.Payload.Content
+
+	// core sign process, generate signature of COSE envelope
+	if err := msg.Sign(rand.Reader, nil, signer); err != nil {
+		return nil, &signature.InvalidSignatureError{Msg: err.Error()}
+	}
+
+	// generate unprotected headers of COSE envelope
+	generateUnprotectedHeaders(req, signer, msg.Headers.Unprotected)
+
+	// TODO: needs to add headerKeyTimeStampSignature here.
+
+	// encode Sign1Message into COSE_Sign1_Tagged object
+	encoded, err := msg.MarshalCBOR()
+	if err != nil {
+		return nil, &signature.InvalidSignatureError{Msg: err.Error()}
+	}
+	e.base = msg
+	return encoded, nil
+}
+
+// Verify implements signature.Envelope interface.
+// Note: Verfiy only verifies integrity of the given COSE envelope
+func (e *envelope) Verify() (*signature.EnvelopeContent, error) {
+	// sanity check
+	if e.base == nil {
+		return nil, &signature.InvalidSignatureError{Msg: "missing COSE signature envelope"}
+	}
+
+	certs, ok := e.base.Headers.Unprotected[cose.HeaderLabelX5Chain].([]interface{})
+	if !ok || len(certs) == 0 {
+		return nil, &signature.InvalidSignatureError{Msg: "COSE envelope invalid certificate chain"}
```

Comment:
We should keep error message consistent with JWS.

https://github.com/notaryproject/notation-core-go/blob/main/signature/jws/envelope.go#L96
```suggestion
		return nil, &signature.InvalidSignatureError{Msg: "certificate chain is not present"}
```